bip-0000.mediawiki

```diff
@@ -0,0 +1,222 @@
+
+
+
+=== Address encoding ===
+
+A silent payment address is constructed in the following manner:
+
+* Let ''B<sub>scan</sub>, b<sub>scan</sub> = Recipient’s [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki BIP340] scan public key and corresponding private key''
+* Let ''B<sub>spend</sub>, b<sub>spend</sub> = Recipient’s [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki BIP340] spend public key and corresponding private key''
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G'', where ''m'' an optional integer tweak for labeling
+** In the case of ''m'' = 0, no label is applied and ''B<sub>m</sub> = B<sub>spend</sub>''
+* The final address is a [https://github.com/bitcoin/bips/bip-0350.mediawkik Bech32m] encoding of:
+** The human-readable part "sp" for mainnet, and "tsp" for testnet
+** The data-part values:
+*** The character "q", to represent a silent payment address of version 0
+*** The 64 byte concatenation of the Recipient's public keys, ''B<sub>m</sub> || B<sub>scan</sub>''
+
+=== Outpoint hash ===
+
+The sender and receiver MUST calculate an outpoint hash for the transaction in the following manner:
+
+* Collect each ''outpoint'' from the transaction, where an outpoint is the `uint256` txid and `uint32` vout pair representing the prevout being spent
+* Sort the outpoints by txid and vout, in ascending order
+* Let ''outpoints = txid<sub>0</sub> || ser<sub>32</sub>(vout<sub>0</sub>) || … || txid<sub>n</sub> || ser<sub>32</sub>(vout<sub>n</sub>)''
+* Let ''outpoint_hash = SHA256(outpoints), truncated to the first 16 bytes''<ref name="why_16byte_hash">'''How was a 16 byte hash chosen?''' In order to support light clients, we expect a full node to calculate ''outpoint_hash·A'', where ''A'' is the sum of the input public keys for a transaction. Using a 16 byte hash instead of a 32 byte will speed up the EC multiplication by ~2x, assuming an optimized implementation. This is in keeping with a server-client model, where we want to minimize the work done by the server. Conversly, we don't want the hash to be too small as this increases the risk of a collision when spending UTXOs from the same public key. Using a 16 byte hash means we can spend from the same public key 6.1×10<sup>6</sup> times with only a ''p = 10<sup>−6</sup>'' probability of a collision (see [https://en.wikipedia.org/wiki/Birthday_problem#Probability_table Birthday Problem] for more details)</ref>
+
+=== Sender ===
+
+==== Selecting inputs ====
+
+The sending wallet performs coin selection as usual with the added restriction that the sending wallet must be able to produce a signature for each UTXO. This excludes any UTXOs not requiring a signature for spending as these do not represent elliptic curve points. Given that these types of scripts are not widely used, no changes to coin selection will be required for the majority of wallets.<ref name="selecting_inputs">'''How to handle multisig and complex scripts''' TODO</ref>
```

Comment:
1) For each UTXO, or for each public key appearing in the output script associated with each UTXO?
2) How does this interact with multisig outputs?
3) I thought the construction was compatible with coinjoins? How is that possible when the sender must know the private keys for all inputs?

Reply:
1. For each public key in the output script, which has a signature
2. For multisig, only the keys which were signed for are used
3. It is but at the cost of increased interactivity between participants

To shed more light on the coinjoin topic: using a single input (vs all) would reveal to the receiver which input was owned by the sender in a coinjoin. When using all the inputs, the receiver can no longer determine which input was owned by the sender, but now the participants in the coinjoin need to cooperate when creating the silent payment address. @RubenSomsen has a separate write-up on this, ill make sure we spell this out in the rationale section

Reply:
Okay sounds good. I figured I’d leave my questions although I already had sort of pieced some things together to document what might benefit from a bit more explanation.